pkg/manager/member/pump_scaler.go

```diff
@@ -0,0 +1,282 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/util"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+)
+
+type pumpScaler struct {
+	generalScaler
+}
+
+// NewPumpScaler returns a pump Scaler
+func NewPumpScaler(deps *controller.Dependencies) *pumpScaler {
+	return &pumpScaler{generalScaler: generalScaler{deps: deps}}
+}
+
+func (s *pumpScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	scaling, _, _, _ := scaleOne(oldSet, newSet)
+	if scaling > 0 {
+		return s.ScaleOut(meta, oldSet, newSet)
+	} else if scaling < 0 {
+		return s.ScaleIn(meta, oldSet, newSet)
+	}
+
+	return nil
+}
+
+func (s *pumpScaler) ScaleOut(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
+	resetReplicas(newSet, oldSet)
+	obj, ok := meta.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("cluster[%s/%s] can't convert to runtime.Object", meta.GetNamespace(), meta.GetName())
+	}
+
+	klog.Infof("scaling out pump statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	var pvcName string
+	switch meta.(type) {
+	case *v1alpha1.TidbCluster:
+		pvcName = fmt.Sprintf("data-%s-pump-%d", meta.GetName(), ordinal)
+	default:
+		return fmt.Errorf("pump.ScaleOut, failed to convert cluster %s/%s", meta.GetNamespace(), meta.GetName())
+	}
+	_, err := s.deps.PVCLister.PersistentVolumeClaims(meta.GetNamespace()).Get(pvcName)
+	if err == nil {
+		_, err = s.deleteDeferDeletingPVC(obj, v1alpha1.PumpMemberType, ordinal)
+		if err != nil {
+			return err
+		}
+		return controller.RequeueErrorf("pump.ScaleOut, cluster %s/%s ready to scale out, wait for next round", meta.GetNamespace(), meta.GetName())
+	} else if !errors.IsNotFound(err) {
+		return fmt.Errorf("pump.ScaleOut, cluster %s/%s failed to fetch pvc informaiton, err:%v", meta.GetNamespace(), meta.GetName(), err)
+	}
+	setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
+	return nil
+}
+
+// parse the advertise address from pod start command.
+// must respect to pump start script in template.go
+func pumpAdvertiseAddr(pod *v1.Pod) string {
```

Comment:
Actually we can construct the advertise addr by following the same rule in the template.go instead of parsing the address from the commands, right?

Reply:
yes, but the addr must exactly match the running one, so if you change the rule, it's hard to make sure you get the right one.